cranelift/codegen/src/egraph/node.rs

```diff
@@ -0,0 +1,382 @@
+        /// identical except for the specific instance will dedup on
+        /// top of each other.
+        inst: Inst,
+        /// Source location, for traps. Not included in Eq/Hash.
+        srcloc: RelSourceLoc,
+    },
+}
+
+impl Node {
+    pub(crate) fn is_non_pure(&self) -> bool {
+        match self {
+            Node::Inst { .. } | Node::Load { .. } => true,
+            _ => false,
+        }
+    }
+}
+
+/// Shared pools for type and id lists in nodes.
+pub struct NodeCtx {
+    /// Arena for result-type arrays.
+    pub types: BumpArena<Type>,
+    /// Arena for arg eclass-ID lists.
+    pub args: ListPool<Id>,
+}
+
+impl std::default::Default for NodeCtx {
+    fn default() -> Self {
+        Self {
+            types: BumpArena::default(),
+            args: ListPool::new(),
+        }
+    }
+}
+
+impl NodeCtx {
+    pub(crate) fn with_capacity(types: usize, args: usize) -> Self {
+        Self {
+            types: BumpArena::arena_with_capacity(types),
+            args: ListPool::with_capacity(args),
+        }
+    }
+}
+
+impl NodeCtx {
+    fn id_eq(&self, a: Id, b: Id, uf: &mut UnionFind) -> bool {
+        uf.find_and_update(a) == uf.find_and_update(b)
+    }
+
+    fn ids_eq(&self, a: &EntityList<Id>, b: &EntityList<Id>, uf: &mut UnionFind) -> bool {
+        let a = a.as_slice(&self.args);
+        let b = b.as_slice(&self.args);
+        a.len() == b.len() && a.iter().zip(b.iter()).all(|(&a, &b)| self.id_eq(a, b, uf))
+    }
+
+    fn hash_id<H: Hasher>(&self, a: Id, hash: &mut H, uf: &mut UnionFind) {
+        let id = uf.find_and_update(a);
+        id.hash(hash);
+    }
+
+    fn hash_ids<H: Hasher>(&self, a: &EntityList<Id>, hash: &mut H, uf: &mut UnionFind) {
+        let a = a.as_slice(&self.args);
+        for &id in a {
+            self.hash_id(id, hash, uf);
+        }
+    }
+}
+
+impl CtxEq<Node, Node> for NodeCtx {
+    fn ctx_eq(&self, a: &Node, b: &Node, uf: &mut UnionFind) -> bool {
+        match (a, b) {
+            (
+                &Node::Param {
+                    block,
+                    index,
+                    ty,
+                    loop_level: _,
+                },
+                &Node::Param {
+                    block: other_block,
+                    index: other_index,
+                    ty: other_ty,
+                    loop_level: _,
+                },
+            ) => block == other_block && index == other_index && ty == other_ty,
+            (
+                &Node::Result { value, result, ty },
+                &Node::Result {
+                    value: other_value,
+                    result: other_result,
+                    ty: other_ty,
+                },
+            ) => self.id_eq(value, other_value, uf) && result == other_result && ty == other_ty,
+            (
+                &Node::Pure {
+                    ref op,
+                    ref args,
+                    ref types,
+                },
```

Comment:
I'm not a huge fan of the pattern of stripping off a layer of reference by sticking an `&` at the beginning of a pattern, and then putting it back with the `ref` keyword. I rather like `Node::Pure { op, args, types }` here, for instance. But I don't feel anywhere near strong enough to, y'know, fight about it or whatever.

Reply:
So in isolation I'd definitely agree! However I tend to like the explicit match-the-borrow form in other cases, so we don't get `Copy` types that nonetheless have to be dereferenced. E.g. elsewhere there is a `&Node::Inst { inst, .. }` -- if we instead did `Node::Inst { inst, .. }` then we'd need `*inst` everywhere. Here I'm trying to remain consistent with that form elsewhere.

(I also think that really confusing errors can arise from this bit of automatic/ergonomic magic and I tend to fall back to pre-2018 Rust out of habit as a result, but that's another discussion :-) )